### PR TITLE
Implement attribute input types

### LINKED ISSUE #4244: Backend changes for storing attribute's input type
There are a couple of things required to store an attribute’s input type.

1. Extend the `Attribute` model with a field to store information about what input type should be used for it. The suggested field name in the Attribute model: input_type. Input types to support for now:
   - Text Field
   - Dropdown
   - MultiSelect
   - Numeric values

2. Check if `AttributeValue` model requires any changes in order to support different types of values.

3. Extend the GraphQL API queries and mutations to support fetching and updating the input_type. For available input types, we should use an enum.

4. The new field(s) should only be available for staff users.

### PR DESCRIPTION
**Pull request by NyanKiyoshi**
Closes #4244.

### Pull Request Checklist

1. [x] Privileged views and APIs are guarded by proper permission checks.
1. [x] All visible strings are translated with proper context.
1. [x] All data-formatting is locale-aware (dates, numbers, and so on).
1. [x] Database queries are optimized and the number of queries is constant.
1. [x] Database migration files are up to date.
1. [x] The changes are tested.
1. [x] The code is documented (docstrings, project documentation).
1. [x] GraphQL schema and type definitions are up to date.
1. [x] Changes are mentioned in the changelog.

### CODE REVIEW

**Comment by django-queries** [comment]
Here is the report for 4f8321283f487076bbabf9eabdf9501ceafb4bc4 (NyanKiyoshi/saleor @ attributes/types)
Base comparison is c4a8e66db69e3e960a95bee958b6c774b9599609.



<details><summary>No differences were found. (click me)</summary>
<p>

```diff
# api.benchmark checkout
  test name                           	left count 	right count
  ------------------------------------	-----------	-----------
  add billing address to checkout     	         41	         41
  add shipping to checkout            	          7	          7
  checkout payment charge             	         16	         16
  complete checkout                   	          6	          6
  create checkout                     	         45	         45

# api.benchmark homepage
  test name                           	left count 	right count
  ------------------------------------	-----------	-----------
  retrieve main menu                  	          5	          5
  retrieve product list               	          4	          4
  retrieve secondary menu             	          5	          5
  retrieve shop                       	          2	          2

# api.benchmark product
  test name                           	left count 	right count
  ------------------------------------	-----------	-----------
  product details                     	         16	         16

# api.benchmark variant
  test name                           	left count 	right count
  ------------------------------------	-----------	-----------
  retrieve variant list               	         13	         13
```

</p>
</details>

File: saleor/graphql/product/descriptions.py [hunk 1]
```diff
@@ -1,4 +1,5 @@
 class AttributeDescriptions:
+    INPUT_TYPE = "The type of rendering for this attribute and its values."
```
Comment (maarcingebala) [inline saleor/graphql/product/descriptions.py]: I would slightly change it:

```suggestion
    INPUT_TYPE = "The input type to use for entering attribute values in the dashboard."
```

File: docs/architecture/products.rst [hunk 2]
```diff
@@ -15,8 +15,18 @@ Type of cover is the only attribute which creates separate variants in our store
 Class Diagram
 -------------
 
+.. To regenerate this class digram:
+    - brew install graphviz
+    - pip3 install --install-option="--include-path=/usr/local/include/" --install-option="--library-path=/usr/local/lib/" pygraphviz
+    - pip install django-extensions
+    - INSTALLED_APPS = [
```
Comment (korycins) [inline docs/architecture/products.rst]: are these commands bash? If no you could add `settings.` to INSTALLED_APPS.  If yes, you should probably add `export`

File: saleor/graphql/product/utils.py [hunk 3]
```diff
@@ -1,45 +1,54 @@
...
+    attributes_json = defaultdict(list)
 
     values_map = {}
     for attr in attributes_queryset:
         for value in attr.values.all():
             values_map[value.slug] = value.id
 
-    for attribute in attribute_value_input:
-        attr_slug = attribute.get("slug")
+    for attribute_input in attribute_value_input:  # type: dict
```
Comment (korycins) [inline saleor/graphql/product/utils.py]: Why do you type `attribute_value_input` here? Don't you think that a better place for that is function parameters?

File: saleor/settings.py [hunk 4]
```diff
@@ -321,6 +321,8 @@ def get_bool_from_env(name, default_value):
     },
 }
 
+INSTALLED_APPS.append("django_extensions")
```
Comment (korycins) [inline saleor/settings.py]: Is it expected?

File: tests/api/test_product.py [hunk 5]
```diff
@@ -1068,6 +1069,55 @@ def test_update_product(
...
+
+
+def test_update_product_can_only_assign_multiple_values_to_valid_input_types(
+    staff_api_client, product, permission_manage_products, color_attribute
+):
+    """Ensures you cannot assign multiple values to input types
+    that are not multi-select. This also ensures multi-select types
+    can be assigned multiple values as intended."""
+
+    staff_api_client.user.user_permissions.add(permission_manage_products)
```
Comment (korycins) [inline tests/api/test_product.py]: Why not simply:
```
response = staff_api_client.post_graphql(
        query, variables, permissions=[permission_manage_products]
    )
```
?

File: saleor/settings.py (hunk 4 above)
Comment (NyanKiyoshi) [reply saleor/settings.py]: Oopsie. Good catch, thanks!

Comment (maarcingebala) [inline saleor/settings.py]: I think this is an unrelated change.

File: tests/api/test_product.py (hunk 5 above)
Comment (NyanKiyoshi) [reply tests/api/test_product.py]: Because it will do two query which is slowing down tests. There is already a test that is ensuring the permissions are respected for the mutation, so there is no need of verifying N more times.
